src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -25,519 +30,222 @@
-        private final int expectedLength;
-
-        @CompilationFinal private boolean hasAppendedObjectArray = false;
-        @CompilationFinal private boolean hasAppendedIntArray = false;
-        @CompilationFinal private boolean hasAppendedLongArray = false;
-        @CompilationFinal private boolean hasAppendedDoubleArray = false;
-
-        public ObjectArrayBuilderNode(int expectedLength) {
-            this.expectedLength = expectedLength;
-        }
-
-        @Override
-        public Object start() {
-            return new Object[expectedLength];
-        }
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class AppendArrayNode extends ArrayBuilderBaseNode {
 
-        @Override
-        public Object start(int length) {
-            if (length > expectedLength) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                final ArrayBuilderNode newNode = new ObjectArrayBuilderNode(length);
-                replace(newNode, length + " > " + expectedLength + " (expected)");
-                return newNode.start(length);
-            }
-
-            return new Object[expectedLength];
+        public static AppendArrayNode create(RubyContext context) {
+            return AppendArrayNodeGen.create(context);
         }
 
-        @Override
-        public Object ensure(Object store, int length) {
-            if (length > ((Object[]) store).length) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                final ArrayBuilderNode newNode = new ObjectArrayBuilderNode(length);
-                replace(newNode, length + " > " + expectedLength + " (expected)");
-                return newNode.ensure(ArrayUtils.copyOf((Object[]) store, length), length);
-            }
+        private final RubyContext context;
 
-            return store;
+        public AppendArrayNode(RubyContext context) {
+            this.context = context;
         }
 
-        @Override
-        public Object appendArray(Object store, int index, DynamicObject array) {
-            Object otherStore = Layouts.ARRAY.getStore(array);
-
-            if (otherStore == null) {
-                return store;
-            }
-
-            if (hasAppendedObjectArray && otherStore.getClass() == Object[].class) {
-                System.arraycopy(otherStore, 0, store, index, Layouts.ARRAY.getSize(array));
-                return store;
-            }
-
-            if (hasAppendedIntArray && otherStore instanceof int[]) {
-                final Object[] objectStore = (Object[]) store;
-                final int[] otherIntStore = (int[]) otherStore;
+        public abstract Object executeAppend(Object array, int index, DynamicObject value);
 
-                for (int n = 0; n < Layouts.ARRAY.getSize(array); n++) {
-                    objectStore[index + n] = otherIntStore[n];
-                }
+        @Specialization(guards = { "generalized.matchesStore(array)", "otherStrategy.matches(other)" },
```

Comment:
Should this be `arrayStrategy.matchesStore(array)`?
Is this modeled after ArrayAppendManyNode or differently?

Reply:
I did it as `generalized.matchesStore(array)` as this implies that the generalised strategy matches the array strategy and so the data will not need to be copied unless the store needs to increase in size.

Reply:
I think it would be much clearer as the complement case of the guards below:
```
"arrayStrategy.matchesStore(array)", "otherStrategy.matches(other)",
"arrayStrategy == generalized"
```
Otherwise it's unclear what this handles or even if it's correct.